Subject: Check for full disk
Last-Update: 2010-10-11

Raise an error if we couldn't write the crontab to disk.
 
Debian-Bug: http://bugs.debian.org/110612

Acked-by: Christian Kastner <debian@kvr.at>
Index: patched/crontab.c
===================================================================
--- patched.orig/crontab.c	2010-05-06 18:18:39.796561016 +0200
+++ patched/crontab.c	2010-05-06 18:18:40.501560966 +0200
@@ -596,19 +596,17 @@
 	Set_LineNum(1)
 	while (EOF != (ch = get_char(NewCrontab)))
 		putc(ch, tmp);
-	ftruncate(fileno(tmp), ftell(tmp));
-	fflush(tmp);  rewind(tmp);
 
-	if (ferror(tmp)) {
-		fprintf(stderr, "%s: error while writing new crontab to %s\n",
-			ProgramName, tn);
+	if (ferror(tmp) || fflush(tmp) || fsync(fd)) {
+		fprintf(stderr, "%s: %s: %s\n",
+			ProgramName, tn, strerror(errno));
 		fclose(tmp);  unlink(tn);
 		return (-2);
 	}
 
 	/* check the syntax of the file being installed.
 	 */
-
+	rewind(tmp);
 	/* BUG: was reporting errors after the EOF if there were any errors
 	 * in the file proper -- kludged it by stopping after first error.
 	 *		vix 31mar87
@@ -658,6 +656,7 @@
 		return (-2);
 	}
 
+
 	if (fclose(tmp) == EOF) {
 		perror("fclose");
 		unlink(tn);
@@ -671,6 +670,8 @@
 		unlink(tn);
 		return (-2);
 	}
+
+
 	log_it(RealUser, Pid, "REPLACE", User);
 
 	poke_daemon();
